File: src/main/scala/marksync/qiita/QiitaService.scala
package marksync.qiita

import java.io.File

import com.softwaremill.sttp.okhttp.OkHttpSyncBackend
import com.softwaremill.sttp.{Id, SttpBackend, Uri, sttp}
import marksync._
import requests.RequestAuth.Bearer

case class QiitaUser(id: String, items_count: Int)

/**
 * Qiita service class.
 *
 * @param username    Username
 * @param accessToken Access Token
 */
class QiitaService(username: String, accessToken: String) extends Service {
  implicit val backend: SttpBackend[Id, Nothing] = OkHttpSyncBackend()

  protected val ENDPOINT = "https://qiita.com/api/v2"
  protected val HEADERS = Map(
    "Content-Type" -> "application/json"
  )

  override val META_FILENAME = "qiita.yml"

  protected lazy val items: Seq[QiitaItem] = {
    val userResponse = requests.get(s"$ENDPOINT/users/$username")
    assert(userResponse.is2xx)
    val user: QiitaUser = Mapper.readJson(userResponse.text(), classOf[QiitaUser])
    (0 to user.items_count / 100).flatMap { i =>
      val itemsResponse = requests.get(s"$ENDPOINT/items", params = Seq(
        "query" -> s"user:$username",
        "page" -> (i + 1).toString,
        "per_page" -> "100"
      ), auth = Bearer(accessToken))
      assert(itemsResponse.is2xx)
      Mapper.readJson(itemsResponse.text(), classOf[Array[QiitaItem]])
    }
  }

  override def getDocuments: Map[String, QiitaItem] = items.map(item => (item.id.get, item)).toMap

  override def getDocument(id: String): Option[ServiceDocument] = items.find(_.id.get == id)

  override def toServiceDocument(doc: Document, uploader: Option[Uploader]): Option[(QiitaItem, Option[String])] = {
    val metaFile = new File(doc.dir, META_FILENAME)

    if (metaFile.exists()) {
      val itemMeta = Option(Mapper.readYaml(metaFile, classOf[QiitaItemMeta])).getOrElse(QiitaItemMeta())
      if (itemMeta.upload.isDefined && uploader.isDefined) {
        uploader.get.meta = itemMeta.upload.get // set upload meta data
      }
      Some((QiitaItem(
        id = itemMeta.id,
        url = itemMeta.url,
        created_at = itemMeta.created_at,
        updated_at = itemMeta.updated_at,
        tags = itemMeta.tags,
        `private` = itemMeta.`private`,
        body = doc.body,
        title = doc.title
      ), itemMeta.digest))
    } else None
  }

  override def saveMeta(file: File, doc: ServiceDocument, uploader: Option[Uploader]): Unit = {
    val item = doc.asInstanceOf[QiitaItem]

    // write qiita.yml
    Mapper.writeYaml(file, QiitaItemMeta(
      id = item.id,
      url = item.url,
      created_at = item.created_at,
      updated_at = item.updated_at,
      digest = Some(item.getDigest(uploader)),
      tags = item.tags,
      `private` = item.`private`,
      upload = uploader.map(_.meta)
    ))
  }

  override def update(serviceDocument: ServiceDocument, uploader: Option[Uploader]): Option[ServiceDocument] = {
    val item = serviceDocument.asInstanceOf[QiitaItem]
    val fileMap = uploader.map(_.getFileMap).getOrElse(Map[String, String]())
    val data = Mapper.getJson(Map(
      "body" -> item.getBody(fileMap),
      "private" -> item.`private`,
      "tags" -> item.tags,
      "title" -> item.getTitle
    ))
    val updateRequest = sttp.auth.bearer(accessToken).headers(HEADERS).body(data)
    val updateResponse = item.id match {
      case Some(itemId) =>
        // update
        updateRequest.patch(Uri.parse(s"$ENDPOINT/items/$itemId").get).send()
      case None =>
        // create
        updateRequest.post(Uri.parse(s"$ENDPOINT/items").get).send()
    }
    updateResponse.body match {
      case Left(errorMessage) =>
        println(s"${updateResponse.statusText}: $errorMessage")
        None
      case Right(deserializedBody) =>
        Some(Mapper.readJson(deserializedBody, classOf[QiitaItem]))
    }
  }
}

File: src/main/kotlin/marksync/services/qiita/QiitaService.kt
package marksync.services.qiita

import marksync.Document
import marksync.Mapper
import marksync.services.Service
import marksync.services.ServiceDocument
import marksync.uploader.FileInfo
import marksync.uploader.Uploader
import java.io.File

/**
 * Qiita service class.
 *
 * @param username Username
 * @param accessToken Access Token
 * @param uploader Uploader
 */
class QiitaService(
    val username: String,
    accessToken: String,
    uploader: Uploader? = null
) : Service(uploader) {
    private val metaFilename = META_FILENAME
    private val apiClient = QiitaApiClient(accessToken)

    private var items: List<QiitaItem>? = null

    override fun getDocuments(): Map<String, QiitaItem> {
        if (items == null) {
            items = apiClient.getItems(username)
        }
        return items!!.map { it.id!! to it }.toMap()
    }

    override fun getDocument(id: String): ServiceDocument? =
        items?.find { it.id == id } ?: apiClient.getItem(id)

    override fun toServiceDocument(doc: Document, dir: File): Pair<QiitaItem, String?>? {
        val metaFile = File(dir, metaFilename)

        return if (metaFile.exists()) {
            val itemMeta = Mapper.readYaml(metaFile, QiitaItemMeta::class.java)
            Pair(
                QiitaItem(
                    id = itemMeta.id,
                    url = itemMeta.url,
                    created_at = itemMeta.created_at,
                    updated_at = itemMeta.updated_at,
                    tags = itemMeta.tags,
                    `private` = itemMeta.`private`,
                    body = doc.body,
                    title = doc.title,
                    files = itemMeta.files
                ),
                itemMeta.digest
            )
        } else null
    }

    override fun createMeta(dir: File) {
        val metaFile = File(dir, metaFilename)
        if (metaFile.exists()) {
            println("${metaFile.name} already exists.")
            return
        }

        // create new marksync.yml
        Mapper.writeYaml(metaFile, QiitaItemMeta())
        println("${metaFile.name} created.")
    }

    override fun saveMeta(doc: ServiceDocument, dir: File, files: ArrayList<FileInfo>) {
        val item = doc as QiitaItem
        val metaFile = File(dir, metaFilename)

        // update marksync.yml
        Mapper.writeYaml(
            metaFile, QiitaItemMeta(
                id = item.id,
                url = item.url,
                created_at = item.created_at,
                updated_at = item.updated_at,
                digest = item.getDigest(),
                tags = item.tags,
                `private` = item.`private`,
                files = files
            )
        )
    }

    override fun update(doc: ServiceDocument): ServiceDocument? {
        val item = doc as QiitaItem
        val data = Mapper.getJson(
            mapOf(
                "body" to item.getDocumentBody(),
                "private" to item.`private`,
                "tags" to item.tags,
                "title" to item.getDocumentTitle()
            )
        )
        return item.id?.let { id ->
            apiClient.updateItem(id, data)
        } ?: apiClient.createItem(data)
    }

    companion object {
        const val META_FILENAME = "marksync.qiita.yml"
    }
}
